How to deal with failing node drains during maintenance (#154)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/update_maintenance.adoc

----
====

. The maintenance of the cluster is only finished once all nodes, including all worker nodes have been upgraded and all `MachineConfigPools` show `Updated=True`.
+
[IMPORTANT]
====
Never leave the cluster in a state with pending node upgrades.

> If the Machine Config operator can't drain a node (eg. because of violating PodDisruptionBudgets) you may have to manually force-drain a node or even manually delete pods that block the node drain.

Reviewer: I'm surprised Vale didn't complain about the `eg.` but that's possibly because that should be spelled `e.g.`.

```suggestion
If the Machine Config operator can't drain a node (for example because doing so would violate a `PodDisruptionBudget`) you may have to manually force-drain a node or even manually delete pods that block the node drain.
```

----
====

> . The maintenance of the cluster is only finished once all nodes, including all worker nodes have been upgraded and all `MachineConfigPools` show `Updated=True`. Never leave the cluster in a state with pending node upgrades. If the Machine Config operator can't drain a node (eg. because of violating PodDisruptionBudgets) you may have to manually force-drain a node or even manually delete pods that block the node drain. Always open a follow-up ticket to investigate the underlying issues if manual intervention is required.

Reviewer: Please keep documentation sources to one sentence per line.

Also, I'd move this section to the `IMPORTANT` box just above it.

Author: In essence, this was already covered in the box above imho. I just want to make it extra obvious what's required for the maintenance to be considered finished, because we had several cases of pending node upgrades after the maintenance in the last weeks.

  logs -c machine-config-daemon -f "${POD}"
----

* If the Machine Config operator fails to drain a node, you may have to force-drain the node:
+
[source,console]
----
oc adm drain <node-name> --delete-emptydir-data --ignore-daemonsets --force --grace-period=0
----
+

> If manually force-draining the node isn't successfull, check which pods are still running on the node with `oc describe node <node-name>` and force delete the pods. The Machine Config operator should then be able to continue with the node upgrades. Depending on what's blocking the drain, these steps may have to be repeated for several nodes.

Reviewer: One sentence per line.

I'd propose to simply use `oc --as=cluster-admin adm drain --ignore-daemonsets --delete-emptydir-data --force --grace-period=0` to force drain nodes instead of manually finding stuck pods and force deleting them.

Author: From experience: If a PDB prevents a pod from being deleted, the drain won't work even with `--force`, only manually deleting the pod allows the operator to continue

Reviewer: Ah, good point, I forgot about that.

  logs -c machine-config-daemon -f "${POD}"
----

* If the Machine Config operator fails to drain a node, you may have to force-drain the node:
+
[source,console]
----
oc --as=cluster-admin adm drain <node-name> --delete-emptydir-data --ignore-daemonsets --force --grace-period=0
----
+

> If manually force-draining the node isn't successfull, check which pods are still running on the node with `oc describe node <node-name>` and force delete the pods.

Reviewer: ```suggestion
If manually force-draining the node isn't successful, check which pods are still running on the node with `oc describe node <node-name>` or `oc get pods --all-namespaces --field-selector spec.nodeName=<node-name>` and force delete any non-daemonset pods shown in the output.
```
